cmd/ltkeycloak/main.go

```diff
@@ -0,0 +1,468 @@
+		"master", // TODO: Allow specifying the master realm
+	)
+	if err != nil {
+		return fmt.Errorf("failed to login to keycloak: %w", err)
+	}
+
+	if !dryRun {
+		// Create group for migrated users if it does not exist
+		_, err := keycloakClient.CreateGroup(ctx, token.AccessToken, deploymentConfig.ExternalAuthProviderSettings.KeycloakRealmName, gocloak.Group{
+			Name: gocloak.StringP(keycloakMigratedGroup),
+		})
+		if err != nil {
+			// Ignore if group already exists
+			if apiErr, ok := err.(*gocloak.APIError); ok && apiErr.Code != 409 {
+				return fmt.Errorf("failed to create group for user migrations in keycloak: %w", err)
+			}
+		}
+	}
+
+	start := 0
+	perPage := 100
+	for {
+		requestCtx, cancel := context.WithTimeout(context.Background(), operationTimeout)
+		defer cancel()
+		users, err := keycloakClient.GetUsers(requestCtx, token.AccessToken, deploymentConfig.ExternalAuthProviderSettings.KeycloakRealmName, gocloak.GetUsersParams{
+			First: &start,
+			Max:   &perPage,
+		})
+		if err != nil {
+			return fmt.Errorf("failed to get users from keycloak: %w", err)
+		}
+
+		for _, user := range users {
+			// Check if user is already migrated
+			if user.Groups != nil && slices.Contains(*user.Groups, keycloakMigratedGroup) {
+				continue
+			}
+
+			// Check if user already exists in Mattermost
+			mmUser, resp, err := mmClient.GetUserByUsername(requestCtx, *user.Username, "")
+			if err != nil && resp.StatusCode != 404 {
+				return fmt.Errorf("failed to get user from mattermost: %w", err)
+			}
+
+			// If user exists in Mattermost and we are not updating existing users, skip
+			if mmUser != nil && !updateExistingUsers {
+				continue
+			}
+
+			if !dryRun && false {
+				if mmUser == nil {
+					// If user does not exist in Mattermost, create it
+					_, _, err = mmClient.CreateUser(requestCtx, &model.User{
+						Username:    *user.Username,
+						Email:       *user.Email,
+						Password:    userPassword,
+						AuthService: model.UserAuthServiceSaml,
+						AuthData:    user.ID,
+					})
+					if err != nil {
+						return fmt.Errorf("failed to create user in mattermost: %w", err)
+					}
+				} else {
+					// If user exists in Mattermost, update it with the new auth data
+					mmUser.AuthData = user.ID
+					mmUser.AuthService = model.UserAuthServiceSaml
+					mmUser.Password = ""
+					_, _, err = mmClient.UpdateUser(requestCtx, mmUser)
+					if err != nil {
+						return fmt.Errorf("failed to update user in mattermost: %w", err)
+					}
+				}
+
+				// Add user to migrated group in keycloak to avoid syncing them again
+				if err := keycloakClient.AddUserToGroup(ctx, token.AccessToken, deploymentConfig.ExternalAuthProviderSettings.KeycloakRealmName, *user.ID, keycloakMigratedGroup); err != nil {
+					return fmt.Errorf("failed to mark user migrated in keycloak: %w", err)
+				}
+			}
+
+			slog.Info("migrated user", slog.String("username", *user.Username))
+
+			if len(users) == 0 {
```

Comment:
Shouldn't this be `len(users) < perPage`? Also, shouldn't this be outside this loop and just below L130?